src/Compilers/CSharp/Portable/Symbols/Source/SourceComplexParameterSymbol.cs

```diff
@@ -127,48 +145,53 @@ private bool HasCallerMemberNameAttribute
                                                   && !HasCallerFilePathAttribute
                                                   && HasCallerMemberNameAttribute;
 
-        private ConstantValue DefaultSyntaxValue
+        private BoundExpression DefaultExpression => DefaultValue.DefaultExpression;
+
+        private ConstantValue DefaultSyntaxValue => DefaultValue.DefaultSyntaxValue;
+
+        private LazyDefaultValue DefaultValue
         {
             get
             {
-                if (_lazyDefaultSyntaxValue == ConstantValue.Unset)
+                if (_lazyDefaultValue == null)
                 {
                     var diagnostics = DiagnosticBag.GetInstance();
 
                     if (Interlocked.CompareExchange(
-                        ref _lazyDefaultSyntaxValue,
+                        ref _lazyDefaultValue,
                         MakeDefaultExpression(diagnostics, ParameterBinderOpt),
-                        ConstantValue.Unset) == ConstantValue.Unset)
+                        null) == null)
                     {
                         // This is a race condition where the thread that loses
                         // the compare exchange tries to access the diagnostics
                         // before the thread which won the race finishes adding
                         // them. https://github.com/dotnet/roslyn/issues/17243
+                        // Same with hasErrors
                         AddDeclarationDiagnostics(diagnostics);
                     }
 
                     diagnostics.Free();
                 }
 
-                return _lazyDefaultSyntaxValue;
+                return _lazyDefaultValue;
             }
         }
 
         // If binder is null, then get it from the compilation. Otherwise use the provided binder.
         // Don't always get it from the compilation because we might be in a speculative context (local function parameter),
         // in which case the declaring compilation is the wrong one.
-        protected ConstantValue MakeDefaultExpression(DiagnosticBag diagnostics, Binder binder)
+        private LazyDefaultValue MakeDefaultExpression(DiagnosticBag diagnostics, Binder binder)
         {
             var parameterSyntax = this.CSharpSyntaxNode;
             if (parameterSyntax == null)
             {
-                return ConstantValue.NotAvailable;
+                return new LazyDefaultValue(null, ConstantValue.NotAvailable);
```

Comment:
Instead of having a constructor that takes null consider adding an overload that takes only a `ConstantValue` parameter.  This helps reduce confusion around where null is / isn't legal.